File: random/JavaScript/divisionEntera.js
/**
 * Divicion de entera lineal
 * @param {number} dividendo 
 * @param {number} divisor 
 * @returns {object} devuelve el cociente y el resto de la divicion
 */
const divisionEntera = (dividendo, divisor) => {
    // time complex O(n)
    //{P: dividendo >= 0 && divisor > 0}
    let cociente = 0, //veamos que el menor numero posible del cociente es 0
        resto = dividendo; // si cociente es 0 => resto = rdividendo segun I
    //{I: dividendo =  cociente * divisor + resto && 0 <= r}(invariante)
    while (resto >= divisor) {
        cociente++;
        resto -= divisor;
    }
    //{Q: dividendo =  cociente * divisor + resto && 0 <= r && resto < divisor}
    return {cociente, resto}
}

/**
 * Divicion de entera mejorada
 * @param {number} dividendo 
 * @param {number} divisor 
 * @returns {object} devuelve el cociente y el resto de la divicion
 */
const divisionEntera2 = (dividendo, divisor) => {
    //el incremento del cociente sera mayor que lineal
    //para eso en vez de tener cociente = cociente + 1
    //vamos a tener cociente = cociente + d (sea d entero y d<=0 )
    // vamos a asumir que d es alguna potencia de 2, por lo tanto va a crecer multiplicandose por 2
    //y el resto (ligado al cociente) en vez de resto = resto - divisor
    //sera resto = resto - d * divisor (esto esta demostrado por calculo)
    // y llamaresmo dd = d * divisor

    //{P: dividendo >= 0 && divisor > 0}
    let cociente = 0,
        resto = dividendo;
    //{I: dividendo =  cociente * divisor + resto && 0 <= r}(invariante)
    while (resto >= divisor) {
        let d = 1, //veamos que 1 es el minimo necesario para que se cumpla II
        dd = divisor; // = (d * divisor) && d = 1 => (1 * divisor) = divisor

        //{II: 1 <= d && d * y <= r && dd = d * y && [∃i:0<=i: 2^i = d]}
        while (resto >= 2 * d * divisor) {
            d *= 2,
            dd *= 2;
        }
        //{QQ: 1 <= d && d * y <= r && dd = d * y && [∃i:0<=i: 2^i = d] && r < 2 * d * y}
        
        cociente += d;
        resto -= dd;
    }
    //{Q: dividendo =  cociente * divisor + resto && 0 <= r && resto < divisor}
    return {cociente, resto}
}

File: random/c/divisionEntera.c
#include <stdio.h>

typedef struct {
    int cociente;
    int resto;
} valoresDivision;

valoresDivision divisionEntera( int dividendo, int divisor ) {
    //{P: dividendo >= 0 && divisor > 0}
    valoresDivision v;
    v.cociente = 0;
    v.resto = dividendo;
    int i = 1;
    //{I: dividendo =  cociente * divisor + resto && 0 <= r}
    while (v.resto >= divisor) {
        v.cociente++;
        v.resto -= divisor;
        printf("Iteracion %d\n", i++);
    }
    //{Q: dividendo =  cociente * divisor + resto && 0 <= r && resto < divisor}
    return v;
}

valoresDivision divisionEnteraMejorada( int dividendo, int divisor ) {
    //{P: dividendo >= 0 && divisor > 0}
    valoresDivision v;
    v.cociente = 0,
    v.resto = dividendo;
    int i = 1;
    //{I: dividendo =  cociente * divisor + resto && 0 <= r}
    while (v.resto >= divisor) {
        int d = 1,
            dd = divisor;
        int j = 1;

        printf("Iteracion %d\n", i++);
        
        while (v.resto >= 2*dd) 
            d *= 2,
            dd *= 2,
            printf("     sub-teracion %d\n", j++);

        v.cociente += d;
        v.resto -= dd; 
    }
    //{Q: dividendo =  cociente * divisor + resto && 0 <= r && resto < divisor}
    return v;
}


int main() {
    int X, Y;

    printf("dividendo: ");
    scanf("%d", &X);
    printf("divisor: ");
    scanf("%d", &Y);

    valoresDivision valores = divisionEntera(X, Y);

    printf("Division Entera\n");
    printf("El cociente es: %d\n", valores.cociente);
    printf("El resto es: %d\n", valores.resto);
    printf("\n");

    valoresDivision valores2 = divisionEnteraMejorada(X, Y);

    printf("Division Entera Mejorada\n");
    printf("El cociente es: %d\n", valores2.cociente);
    printf("El resto es: %d\n", valores2.resto);
    printf("\n");


    return 0; 
}